### PR TITLE
Multiple fixes for getting pub key from legacy DH PKEY

### LINKED ISSUE #16247: Cannot fetch OSSL_PKEY_PARAM_PUB_KEY from legacy DH key
Fetching OSSL_PKEY_PARAM_PUB_KEY from a legacy DH does not work. The assertion in the following code fails:

```
#include <assert.h>
#include <openssl/core_names.h>
#include <openssl/dh.h>
#include <openssl/evp.h>

int main() {
	EVP_PKEY *key = EVP_PKEY_new();
	DH *dh = DH_new();
	DH_generate_parameters_ex(dh, 512, DH_GENERATOR_2, NULL);
	DH_generate_key(dh);
	EVP_PKEY_assign_DH(key, dh);

	BIGNUM *bn = NULL;
	EVP_PKEY_get_bn_param(key, OSSL_PKEY_PARAM_PUB_KEY, &bn);
	assert(bn != NULL);
	EVP_PKEY_free(key);
	return 0;
}
```

From some cursory debugging, I believe this is due to a bad interaction between https://github.com/openssl/openssl/blob/46d51520319b8e4220b203c517b1232cf33a579f/crypto/evp/ctrl_params_translate.c#L2252 and 
https://github.com/openssl/openssl/blob/46d51520319b8e4220b203c517b1232cf33a579f/crypto/evp/ctrl_params_translate.c#L659.

### PR DESCRIPTION
**Pull request by t8m**
There were multiple issues with getting OSSL_PKEY_PARAM_PUB_KEY
from a legacy EVP_PKEY DH and DHX keys.

Fixes #16247

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -2481,13 +2481,21 @@ static int test_EVP_PKEY_set1_DH(void)
...
-            || !BN_set_word(g, 2)
+            || !TEST_ptr(pubkey = BN_new())
+            || !TEST_true(BN_set_word(p, 9999))
+            || !TEST_true(BN_set_word(g, 2))
+            || !TEST_true(BN_set_word(pubkey, 4321))
             || !TEST_ptr(noqdh = DH_new())
-            || !DH_set0_pqg(noqdh, p, NULL, g))
+            || !TEST_true(DH_set0_pqg(noqdh, p, NULL, g))
+            || !TEST_true(DH_set0_key(noqdh, pubkey, NULL))
+            || !TEST_ptr(pubkey = BN_new())
```
Comment (slontis) [inline test/evp_extra_test.c:2497]: This looks like it leaks pubkey with a double new

Comment (t8m) [reply test/evp_extra_test.c:2497]: No, the previous pubkey is set0 into the noqdh.

**Comment by t8m** [comment]
Ping for review

**Comment by beldmit** [comment]
The changes look reasonably to me, though I don't understand some of them

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the review.

**Pull request closed without merging**